archey/entries/uptime.py

```diff
@@ -8,41 +17,147 @@ class Uptime(Entry):
                     uptime += ' and '
-
                 else:
                     uptime += ', '
 
-        if hour:
-            uptime += str(hour) + ' hour'
-            if hour > 1:
+        if hours:
+            uptime += str(hours) + ' hour'
+            if hours > 1:
                 uptime += 's'
 
-            if minute:
+            if minutes:
                 uptime += ' and '
 
-        if minute:
-            uptime += str(minute) + ' minute'
-            if minute > 1:
+        if minutes:
+            uptime += str(minutes) + ' minute'
+            if minutes > 1:
                 uptime += 's'
-
-        else:
-            if not day and not hour:
-                uptime = '< 1 minute'
+        elif not days and not hours:
+            uptime = '< 1 minute'
 
         self.value = uptime
+
+    def _get_uptime_delta(self):
+        """
+        Returns a `datetime.timedelta` instance containing the machine uptime.
+        Tries a variety of methods, increasing compatibility for a wide range of systems.
+        """
+        # Try the /proc/uptime file
+        try:
+            return self._proc_file_uptime()
+        except (PermissionError, FileNotFoundError):
+            # Can't read /proc/uptime.
+            # Not GNU/Linux ? Limited permissions ?
+            pass
+
+        # Try the python `time` module clocks
+        try:
+            return self._clock_uptime()
+        except RuntimeError:
+            # Probably Python <3.7, or not *NIX
+            pass
+
+        # FUTURE: Windows support could be added here with the `wmi` or `pywin32` module.
+
+        # Fall back to the `uptime` command.
+        return self._parse_uptime_cmd()
+
+    @staticmethod
+    def _proc_file_uptime():
+        """Tries to get uptime using the `/proc/uptime` file"""
+        with open('/proc/uptime') as f_uptime:
+            return timedelta(
+                seconds=float(f_uptime.read().split()[0])
+            )
+
+    @staticmethod
+    def _clock_uptime():
+        """Tries to get uptime using the clocks from the Python `time` module"""
+        # Try: Linux uptime clock, macOS uptime clock, BSD uptime clock.
+        for clock in ('CLOCK_BOOTTIME', 'CLOCK_UPTIME_RAW', 'CLOCK_UPTIME'):
+            try:
+                return timedelta(
+                    seconds=time.clock_gettime(getattr(time, clock))
+                )
+            except AttributeError:
+                pass
+
+        # Probably Python <3.7, or just not one of the above OSes
+        raise RuntimeError
+
+    @staticmethod
+    def _parse_uptime_cmd():
+        """Tries to get uptime by parsing the `uptime` command"""
+        try:
+            uptime_output = check_output('uptime')
+        except FileNotFoundError:
+            # No `uptime` command.
+            # Since `procps` is a dependency (which provides `uptime`) we can just exit here.
+            # Note: We _should_ not got there as `Processes` first check `procps` availability.
```

Comment:
Except if `Processes` is disabled in the configuration...!

Reply:
That's not possible as `Processes` is not an entry but the singleton automatically-populated on startup :wink:

Reply:
Oops! Sorry, I thought it was an entry 😖